cli/src/complete.rs

```diff
@@ -0,0 +1,104 @@
+        .and_then(|stdout| String::from_utf8(stdout).ok())
+        .map(|stdout| stdout.lines().map(CompletionCandidate::new).collect())
+        .unwrap_or_default()
+}
+
+/// Shell out to jj during dynamic completion generation
+///
+/// This is necessary because dynamic completion code needs to be aware of
+/// global configuration like custom storage backends. Dynamic completion
+/// code via clap_complete doesn't accept arguments, so they cannot be passed
+/// that way. Another solution would've been to use global mutable state, to
+/// give completion code access to custom backends. Shelling out was chosen as
+/// the preferred method, because it's more maintainable and the performance
+/// requirements of completions aren't very high.
+fn jj() -> std::process::Command {
+    let mut command = std::process::Command::new(
+        std::env::current_exe().unwrap_or(std::path::PathBuf::from("jj")),
+    );
+
+    // Snapshotting could make completions much slower in some situations
+    // and be undesired by the user.
+    command.arg("--ignore-working-copy");
+
+    // Parse some of the global args we care about for passing along to the
+    // child process. This shouldn't fail, since none of the global args are
+    // required.
+    let app = crate::commands::default_app();
+    let config = config::Config::builder()
+        .add_source(default_config())
+        .build()
+        .unwrap();
+    let mut layered_configs = LayeredConfigs::from_environment(config);
+    let ui = Ui::with_config(&layered_configs.merge()).unwrap();
+    let cwd = std::env::current_dir()
+        .and_then(|cwd| cwd.canonicalize())
+        .unwrap();
+    let maybe_cwd_workspace_loader = DefaultWorkspaceLoaderFactory.create(find_workspace_dir(&cwd));
+    layered_configs.read_user_config().unwrap();
+    if let Ok(loader) = &maybe_cwd_workspace_loader {
+        layered_configs
+            .read_repo_config(loader.repo_path())
+            .unwrap();
+    }
+    let config = layered_configs.merge();
+    // skip 2 because of the clap_complete prelude: jj -- jj <actual args...>
+    let args = std::env::args().skip(2).collect();
+    let args = resolve_default_command(&ui, &config, &app, args).unwrap();
+    let args = resolve_aliases(&ui, &config, &app, args).unwrap();
+    let args = app
+        .clone()
+        .disable_version_flag(true)
+        .disable_help_flag(true)
+        .disable_help_subcommand(true)
+        .ignore_errors(true)
+        .get_matches_from(args);
+    let args: GlobalArgs = GlobalArgs::from_arg_matches(&args).unwrap();
+
+    if let Some(repository) = args.repository {
+        command.arg("--repository");
+        command.arg(repository);
+    }
+    if let Some(at_operation) = args.at_operation {
+        command.arg("--at-operation");
+        command.arg(at_operation);
+    }
+    for config_toml in args.early_args.config_toml {
+        command.arg("--config-toml");
+        command.arg(config_toml);
+    }
```

Comment:
I believe these should be the only global args relevant for completions. I only added a test for the `--repository` flag for now, since the other two are not relevant for renaming bookmarks.